packages/nvidia_nat_mcp/src/nat/plugins/mcp/client_impl.py

```diff
@@ -75,103 +87,217 @@ def mcp_client_transport(self, transport: str | None):
+                    logger.info(" Total sessions: %d", len(self._session_clients))
+                except Exception as e:
+                    logger.warning("Error cleaning up session client %s: %s", truncate_session_id(session_id), e)
+
+    async def _get_session_client(self, session_id: str) -> MCPBaseClient:
+        """Get the appropriate MCP client for the session."""
+        # Throttled cleanup on access
+        now = datetime.now()
+        if now - self._last_cleanup_check > self._cleanup_check_interval:
+            await self._cleanup_inactive_sessions()
+            self._last_cleanup_check = now
+
+        # If the session_id equals the configured default_user_id use the base client
+        # instead of creating a per-session client
+        default_uid = self._shared_auth_provider.config.default_user_id
+        if default_uid and session_id == default_uid:
+            return self.mcp_client
+
+        # Fast path: check if session already exists (no lock needed)
+        if session_id in self._session_clients:
+            # Update last activity for existing client
+            self._session_last_activity[session_id] = datetime.now()
+            return self._session_clients[session_id]
+
+        # Slow path: create session with proper locking
+        async with self._session_creation_lock:
+            # Double-check after acquiring lock (another coroutine might have created it)
+            if session_id in self._session_clients:
+                self._session_last_activity[session_id] = datetime.now()
+                return self._session_clients[session_id]
+
+            # Check session limit before creating new client
+            if len(self._session_clients) >= self._client_config.max_sessions:
+                # Try cleanup first to free up space
+                await self._cleanup_inactive_sessions()
+
+                # Re-check after cleanup
+                if len(self._session_clients) >= self._client_config.max_sessions:
+                    logger.warning("Session limit reached (%d), rejecting new session: %s",
+                                   self._client_config.max_sessions,
+                                   truncate_session_id(session_id))
+                    raise RuntimeError(
+                        f"Service temporarily unavailable: Maximum concurrent sessions ({self._client_config.max_sessions}) exceeded. Please try again later."
+                    )
+
+            # Create session client lazily
+            logger.info("Creating new MCP client for session: %s", truncate_session_id(session_id))
+            session_client = await self._create_session_client(session_id)
+
+            # Cache the client
+            self._session_clients[session_id] = session_client
+            self._session_last_activity[session_id] = datetime.now()
+            return session_client
+
+    @asynccontextmanager
+    async def _session_usage_context(self, session_id: str):
```

Comment:
Here is where I'd use a per-session asyncio.Lock to ensure mutex.

Before grabbing the session lock you'd want to grab the session_rwlock in read mode.

Reply:
done